src/audio/frame.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <QDebug>
+#include <cmath>
+#include <limits>
+
+#include "engine/engine.h"
+#include "util/fpclassify.h"
+
+namespace mixxx {
+namespace audio {
+/// FrameDiff_t can be used to store the difference in position between
+/// two frames and to store the length of a segment of track in terms of frames.
+typedef double FrameDiff_t;
+
+/// FramePos defines the position of a frame in a track
+/// with respect to a fixed origin, i.e. start of the track.
+class FramePos final {
+  public:
+    typedef double value_t;
+    static constexpr value_t kStartValue = 0;
+    static constexpr value_t kInvalidValue = std::numeric_limits<FramePos::value_t>::quiet_NaN();
+
+    constexpr FramePos()
+            : m_framePosition(kInvalidValue) {
+    }
+
+    constexpr explicit FramePos(value_t framePosition)
+            : m_framePosition(framePosition) {
+    }
+
+    static constexpr FramePos fromEngineSamplePos(double engineSamplePos) {
+        return FramePos(engineSamplePos / mixxx::kEngineChannelCount);
+    }
+
+    constexpr double toEngineSamplePos() const {
+        return value() * mixxx::kEngineChannelCount;
+    }
+
+    bool isValid() const {
+        return !util_isnan(m_framePosition);
+    }
+
+    void setValue(value_t framePosition) {
+        m_framePosition = framePosition;
+    }
+
+    constexpr value_t value() const {
+        return m_framePosition;
+    }
+
+    bool isFractional() const {
+        value_t integerPart;
+        return std::modf(value(), &integerPart) != 0;
+    }
+
+    [[nodiscard]] FramePos toFullFrames() const {
+        return FramePos(std::floor(value()));
+    }
+
+    FramePos& operator+=(FrameDiff_t increment) {
+        m_framePosition += increment;
+        return *this;
+    }
+
+    FramePos& operator-=(FrameDiff_t decrement) {
+        m_framePosition -= decrement;
+        return *this;
+    }
+
+    FramePos& operator*=(double multiple) {
+        m_framePosition *= multiple;
+        return *this;
+    }
+
+    FramePos& operator/=(double divisor) {
+        m_framePosition /= divisor;
+        return *this;
+    }
+
+  private:
+    value_t m_framePosition;
+};
+
+/// FramePos can be added to a FrameDiff_t
+inline FramePos operator+(FramePos framePos, FrameDiff_t frameDiff) {
+    return FramePos(framePos.value() + frameDiff);
+}
+
+/// FramePos can be subtracted from a FrameDiff_t
+inline FramePos operator-(FramePos framePos, FrameDiff_t frameDiff) {
+    return FramePos(framePos.value() - frameDiff);
+}
+
+/// Two FramePos can be subtracted to get a FrameDiff_t
+inline FrameDiff_t operator-(FramePos framePos1, FramePos framePos2) {
+    return framePos1.value() - framePos2.value();
+}
+
+// Adding two FramePos is not allowed since every FramePos shares a common
```

Comment:
What do you mean? There is already a compile-time assertion so to say because the operator does not exist. So you just can't do it and get a compiler error.

Reply:
Using the type system is superior to runtime checks. That's the main reason why this wrapper class has been introduced!